### PR TITLE
GridSearchCV and Pipeline

### PR DESCRIPTION
**Pull request by mrocklin**
This moves over a prototype started by @amueller and myself here:  github.com/mrocklin/dask-learn

This still needs quite a bit of cleaning up but at least it's someplace slightly more visible now.

### CODE REVIEW

**Comment by mrocklin** [comment]
@amueller this PR, along with https://github.com/blaze/dask/pull/742 provides the following graph for the grid search + pipeline

gridsearch

This looks better to my eye.  We still need some work to avoid excess tokenization but at least now we're correct.

**Comment by amueller** [comment]
This looks correct at a cursory glance. Well the result was correct, the sharing is now better :) Thanks!
I'm a bit busy right now, not sure when I can work on this again :-/ but I hope soonish.

**Comment by shoyer** [comment]
Just one thought -- this should definitely hook up nicely (aka dask imperative) when provided with input in the form of a dask array or dataframe.

**Comment by mrocklin** [comment]
@gaelvaroquaux wrote the following in https://github.com/mrocklin/dask-learn/issues/2 . Moving that discussion here.

Hi,

I was looking at the code in grid_search.py to understand how things work, and I am really dense: I don't understand where is the dask graph built.

It seems to me that _fit_and_score does not do any delayed evaluation and directly runs the estimation. When I look at line 67, where the for loop is run, the call is not protected. Actually, the usage of dask is tiny in this codebase, pretty much limited to line 75, which seems to me to be an execution of the dask graph. But nothing before seems to built the graph

Thus I do not understand how this work. Is it because dask arrays are passed in, and thus they trace what happens to them in the estimator? In which case I fear that this will work only in a small number of estimators.

**Comment by mrocklin** [comment]
@gaelvaroquaux most of the interest graph construction occurs in pipeline.py.  Gridsearch is just, as you suggest, a for loop.  The main utility of dask is that the graphs produced by `Pipeline` can be merged so that they share structure.  The image a few comments illustrates this.

Please note that this PR is the result of a sprint and very very rough.  Further feedback would be quite helpful before moving forward.

**Comment by GaelVaroquaux** [comment]
OK, now this makes more sense.

So basically, what we have in terms of design is:

- A delayed scorer
- A pipeline that delays operation / builds a graph
- A trivial modification to GridSearch to run the computation from that.

Interesting!

I wonder if the modifications could not be made even lighter by using scikit-learn's unmodified _fit_and_score and, given a scorer modify it to return a dask graph item. Of course, with the current scikit-learn you'll get an error as there is some type-checking done on the return of a scorer, but we could think of loosening this type checking. As it seems possible to integrate something in spirit of the modification to GridSearch, than the only thing to do outside of scikit-learn would be a trivial subclass of GridSearch that:
- checks if the estimator is a pipeline and if so clones it to a dask pipeline
- decorates the scorer

That would basically adapt the GridSearch + Pipeline pattern of scikit-learn to dask while minimizing duplication and most likely being future prone (I am afraid of the tight coupling between the code bases that is likely to break each time that there is a refactoring of the one of them).

**Comment by amueller** [comment]
Yeah it would be great if we could do this without modifying grid-search.
We probably want to modify _fit_and_score so we can delay the slicing of X. Currently, the slicing of X is done before the graph is build basically (because the graph is build in the pipeline.fit).

And the current implementation doesn't know anything about scorers. That should be straight-forward to fix. It just uses the score of the estimators. I wanted to get something running quickly (this was a one-day sprint).

**Comment by amueller** [comment]
@mrocklin is there any way to get rid of line 75? Not really because compute doesn't to into lists, as you said. Would that be possible to change? And would that change help here?

**Comment by amueller** [comment]
The amount of copy and paste here is also a sign that our BaseSearchCV.fit is pretty spaghetti. If we refactor it, we could just inherit from BaseSearchCV and add a line to a much smaller fit, with much less copy and paste.

**Comment by mrocklin** [comment]
If you're trying to remove `value` but not `compute` then yes, that's fine, you can do

    Before: out, = compute(value(out))
    After: out = compute(*out)

However, if you're also trying to remove `compute` then we'll need to make the following code chain lazy-friendly.

**Comment by amueller** [comment]
Well, we can't really make everything lazy-friendly unless we assume that the estimator's fit is lazy.
Also, writing it in such a way that the argmax can be lazy is probably odd.
We need the numpy duck-arrays for that maybe ;)

**Comment by mrocklin** [comment]
I would be very happy if small changes could be made to sklearn that would reduce the amount of sklearn code that needs to be maintained within dask.  

From the perspective of gridsearch I think it would be sufficient to allow a function that takes a sequence of potentially-lazy-results and turns them into a sequence of concrete-results.  Perhaps appropriate sklearn functions/objects take this as function as an input (or as a static method on the class):

```python
def identity(*args):
    return args

def f(*args, reify=identity):
    ...
    out = reify(out)
    ...
```

Although something like this may be overly invasive into the sklearn codebase.

**Comment by GaelVaroquaux** [comment]
> Although something like this may be overly invasive into the sklearn codebase.

Yes :(

**Comment by mrocklin** [comment]
Oh well.

**Comment by mrocklin** [comment]
What is the right path forward on this?

I can continue to refine things here a bit, mostly centered around reducing tokenizing/hashing overhead in pipeline.  I hope to leave the sklearn refactoring to sklearn devs.

**Comment by amueller** [comment]
I think it's mostly up to us now to make this fully featured. It would be nice to make the slicing lazy, not sure that is hard. Probably not. I'll be somewhat offline for the next week, trying to finally get my health in order :-/

**Comment by mrocklin** [comment]
General question: in what contexts is this approach useful outside of `BaseSearchCV` and `Pipeline` interactions?  If it's just these two then I may rewrite this all as a single function.

**Comment by amueller** [comment]
I think it would be good to separate these two.
Calling ``cross_val_score`` on ``Pipeline`` is another use-case.
And in the future it might be interesting to support Something like ``GridSearchCV(RFE(Pipeline(...))``.

**Comment by shoyer** [comment]
There's also `cross_val_predict`, which is one of my favorites.

**Comment by cowlicks** [comment]
How mature should this be before it gets merged? I think it would be better to get a minimal use-case in: gridsearch, and pipeline. Then add these other features later. Or does that work already?

It would be nice to get this in the next release (in a few days?), with a big "experimental" label.

**Comment by mrocklin** [comment]
I think that this is too experimental to be included in the next release.

**Comment by GaelVaroquaux** [comment]
> I think that this is too experimental to be included in the next release.

+1

**Comment by amueller** [comment]
Yeah in particular because it just silently ignores the scoring, I think.

**Comment by mrocklin** [comment]
OK.  I've moved this over to a small project here: https://github.com/mrocklin/dasklearn

I've stopped using dictionaries directly and have gone instead towards using `dask.do.`  I like the result however it also requires a couple of fixes within dask to make pleasant.  Some of those are happening [here](https://github.com/blaze/dask/pull/779).

**Comment by mrocklin** [comment]
The resulting graph of the gridsearch test is here:

dask

Things to do

- [ ] profile and minimize hashing costs 
- [ ] figure out how to properly hash down an estimator https://gitter.im/scikit-learn/scikit-learn?at=5617f5bc99bbd76f0f308f61

cc @jcrist

**Comment by amueller** [comment]
sweet :)
One of us should look at it to complete the api and remove all the hacks I did, like not supporting scorers.

**Comment by mrocklin** [comment]
I'm happen to add collaborators to that repository if anyone is interested or operate under a forking model.  I don't think we'll have many conflicts.  I plan to continue to use this issue for communication.

**Comment by amueller** [comment]
Sounds good :)

**Comment by mrocklin** [comment]
Closing this as an active PR.  Feel free to keep communication going though.
